beacon-chain/p2p/handshake.go

```diff
@@ -54,37 +56,47 @@ func (s *Service) AddConnectionHandler(reqFunc func(ctx context.Context, id peer
 						// used in production.
 						log.Debug("Not disconnecting peer with unsupported protocol. This may be the DHT node or relay.")
 						s.host.ConnManager().Protect(conn.RemotePeer(), "relay/bootnode")
+						s.peers.SetConnectionState(conn.RemotePeer(), peers.PeerDisconnected)
 						return
 					}
+					s.peers.IncrementBadResponses(conn.RemotePeer())
+					badResponses, err := s.peers.BadResponses(conn.RemotePeer())
+					if err == nil && badResponses == s.peers.MaxBadResponses() {
+						log.Info("Peer has given too many bad responses; will ignore in future")
+					}
+					s.peers.SetConnectionState(conn.RemotePeer(), peers.PeerDisconnecting)
 					if err := s.Disconnect(conn.RemotePeer()); err != nil {
-						log.WithError(err).Errorf("Unable to close peer %s", conn.RemotePeer())
-						return
+						log.WithError(err).Error("Unable to disconnect from peer")
 					}
+					s.peers.SetConnectionState(conn.RemotePeer(), peers.PeerDisconnected)
 					return
 				}
-				log.WithField("peer", conn.RemotePeer().Pretty()).Info("New peer connected")
+				s.peers.SetConnectionState(conn.RemotePeer(), peers.PeerConnected)
+				log.WithField("active", len(s.peers.Active())).Info("Peer connected")
 			}()
 		},
 	})
 }
 
-// AddDisconnectionHandler ensures that previously disconnected peers aren't dialed again. Due
-// to either their ports being closed, nodes are no longer active,etc. This also calls the handler
-// responsible for maintaining other parts of the sync or p2p system.
+// AddDisconnectionHandler disconnects from peers.  It handles updating the peer status.
+// This also calls the handler responsible for maintaining other parts of the sync or p2p system.
 func (s *Service) AddDisconnectionHandler(handler func(ctx context.Context, id peer.ID) error) {
 	s.host.Network().Notify(&network.NotifyBundle{
 		DisconnectedF: func(net network.Network, conn network.Conn) {
+			log := log.WithField("peer", conn.RemotePeer().Pretty())
 			// Must be handled in a goroutine as this callback cannot be blocking.
 			go func() {
-				s.exclusionList.Set(conn.RemotePeer().String(), true, 1)
-				log := log.WithField("peer", conn.RemotePeer())
-				log.Debug("Peer is added to exclusion list")
+				s.peers.SetConnectionState(conn.RemotePeer(), peers.PeerDisconnecting)
 				ctx := context.Background()
 				if err := handler(ctx, conn.RemotePeer()); err != nil {
 					log.WithError(err).Error("Failed to handle disconnecting peer")
```

Comment:
in the event of failure, you will have to reverse back the connection state to `peers.PeeConnected`

Reply:
Resolving as per conversation.